flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -151,7 +170,13 @@ private static boolean isPreFlightFailure(final Throwable cause) {
 	 */
 	public interface FailJobCallback {
 
-		void failJob(final Throwable cause);
+		/**
+		 * Fails the whole job graph.
+		 *
+		 * @param cause The reason why the job is cancelled.
+		 * @param failingAttempt The failing attempt id to prevent failing the wrong job.
+		 */
+		void failJob(final Throwable cause, @Nullable final ExecutionAttemptID failingAttempt);
```

Comment:
Can we make `failingAttempt` parameter obligatory? If not, it would be better to avoid null by overloading this method.

Reply:
I'm afraid we cannot since there're cases like exception thrown when `triggerCheckpoint`, as well as the legacy `handleSynchronousSavepointFailure` method in `CheckpointFailureManager`, which are not caused by a single task failure. Will add a default interface to overload this method.